### PR TITLE
[Task 1] - Codebase Review

### PR DESCRIPTION
**Pull request by eridervishi**
## General Project Recommendations

1. Adopt TypeScript for better error handling, static typing, better readability and maintainability

2. Use ESLint and Prettier for consistency across the codebase, catches potential errors early

3. Use RTK Query for consistent caching and state management for server state, error handling, declarative data fetching

4. Use Prop-Types (if not transitioning to TypeScript) to enhance error handling and type checking

5. Ensure reducers control the type of the payload to improve type safety and predictability

6. Implement error handling strategy by using error boundaries and consistent error logging and reporting

7. Optimize performance by implementing code splitting and lazy loading, using memoization techniques (useMemo, useCallback, React.memo), and optimizing re-renders with careful state management    

8. Refactor project folder structure to follow the feature-based approach to improve organization, readability and scalability

9. Enhance accessibility (a11y) throughout the application by ensuring proper ARIA attributes, keyboard navigation, and screen reader support

10. Upgrade to RTK 2.x for better performance, improved readability and maintainability

11. Upgrade to React Router v6 for better consistency with modern React practices, more intuitive APIs
    
### constants.js

1. Security Issue: API_KEY is exposed in the code (line 1)
   - Store in a secure environment variable (e.g., .env file)

2. Use more descriptive names for constants
   - Rename ENDPOINT to BASE_API_URL for clarity

3. Group related constants
   ```javascript
   export const API = {
     BASE_URL: 'https://api.themoviedb.org/3',
     KEY: process.env.REACT_APP_API_KEY,
   };
   
   export const ENDPOINTS = {
     DISCOVER: '/discover/movie',
     SEARCH: '/search/movie',
     MOVIE: '/movie',
   };
   ```

### App.js

1. Simplify getMovies (line 46)
   - Remove the check for searchQuery presence
   - Transfer logic to the reducer
   ```javascript
   const getMovies = () => {
     dispatch(fetchMovies(searchQuery));
   };
   ```

2. Issue with useEffect missing dependency (line 73)
   - transform to customHook
   - If not, use useCallback for getMovies function
   ```javascript
   const getMovies = useCallback(() => {
     dispatch(fetchMovies(searchQuery));
   }, [dispatch, searchQuery]);

   useEffect(() => {
     getMovies();
   }, [getMovies]);
   ```

3. Improve searchMovies function (line 41)
   - Remove unnecessary navigation
   - Use a more descriptive name like handleSearch
   ```javascript
   const handleSearch = (query) => {
     setSearchParams(query ? { search: query } : {});
     dispatch(fetchMovies(query));
   };
   ```

4. Avoid Prop Drilling
   - Connect child components to Store and get their slices

5. Separate concerns
   - Move API calls to a separate service file
   - Create custom hooks for complex logic (e.g., useMovieSearch)

### Movie.jsx

1. Improve card open/close behavior (line 24)
   - Use state to control the card's open/close behavior
   ```javascript
   const [isOpen, setIsOpen] = useState(false);
   
   // In JSX
   <div className={`card ${isOpen ? 'opened' : ''}`} onClick={() => setIsOpen(!isOpen)}>
   ```

2. Consistent payload for star/unstar actions (lines 32, 43)
   - Ensure both actions receive the same payload structure
   ```javascript
   const moviePayload = {
     id: movie.id,
     title: movie.title,
     overview: movie.overview,
     release_date: movie.release_date,
     poster_path: movie.poster_path,
   };
   
   // For both star and unstar actions
   dispatch(starMovie(moviePayload));
   dispatch(unstarMovie(moviePayload));
   ```

3. Use selectors properly to get only the needed data from the store (line 8)

### Header.jsx

1. Fix search input (line 13)
   - Remove it from Link component
   - Implement controlled input with state
   ```javascript
   const [searchTerm, setSearchTerm] = useState('');
   
   const handleSearch = (e) => {
     e.preventDefault();
     searchMovies(searchTerm);
   };
   
   // In JSX
   <form onSubmit={handleSearch}>
     <input
       type="search"
       value={searchTerm}
       onChange={(e) => setSearchTerm(e.target.value)}
       // ... other props
     />
   </form>
   ```

2. Add proper label for accessibility
   ```jsx
   <label htmlFor="movie-search">Search movies</label>
   <input id="movie-search" ... />
   ```

3. Implement debounce for search input to reduce unnecessary API calls

4. Use NavLink consistently for navigation items (line 12)

5. Implement search functionality here

### Redux Files (moviesSlice.js, starredSlice.js, watchLaterSlice.js)

1. Implement proper error handling in async thunks and error messages (moviesSlice.js line 4)

2. Use createSelector for memoized selectors

3. Normalize state shape for better performance with large datasets

### Test Files (*.test.js)

1. Increase test coverage
   - Add more unit tests for individual components
   - Implement integration tests for connected components

2. Use mock service worker (MSW) for mocking API calls

3. Implement snapshot testing for UI components

4. Test error scenarios and edge cases

5. Use React Testing Library for more user-centric tests

### CSS Files (*.scss)

1. Implement CSS Modules or a CSS-in-JS solution for better encapsulation

2. Use a consistent naming convention (e.g., BEM)

3. Implement a theming solution for easier customization

4. Optimize CSS for performance (minimize, remove unused styles)

5. Ensure responsive design across all components

### index.js

1. Implement code splitting and lazy loading

2. Add error boundary at the app root

3. Use React.StrictMode for identifying potential problems

### CODE REVIEW

File: src/constants.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export const API_KEY = '8cac6dec66e09ab439c081b251304443'
```
Comment (eridervishi) [inline src/constants.js:1]: Security Issue: API_KEY is exposed in the code (line 1)
   - Store in a secure environment variable (e.g., .env file)
